Backblaze Storage Driver : Include get_container and get_object methods

jshridha opened the pull request:
Include get_container and  get_object methods in the BackblazeB2StorageDriver

This is required for compatibility with libraries that reference these methods.

@Kami Since you wrote the initial backblaze implementation (I know it's still under dev), it would be great if you could review this. Thanks!

jshridha committed:
Include the get_container, get_object, and upload_object_via_stream methods

tonybaloney commented:
Great contribution. is it possible to remove the dependency on the os stdlib?

tonybaloney commented:
Oh, and please can you develop and add some tests for this as well. We don't have CI tests for Windows with Travis CI so unittests as a minimum.

jshridha commented:
Can do! I just added in some unit tests, got rid of the os and sys dependencies, reformatted according style guide a bit better, and rebased the pull request. Let me know if there's any other comments, I'm happy to make changes.

tonybaloney commented on libcloud/storage/drivers/backblaze_b2.py:
```diff
@@ -292,7 +316,7 @@ def download_object_as_stream(self, obj, chunk_size=None):
 
     def upload_object(self, file_path, container, object_name, extra=None,
-                      verify_hash=True, headers=None):
+                      verify_hash=True, headers=None, iterator=None):
```
@Kami is that ok?

Kami replied:
There is already a method for that - `upload_object_via_stream`. That method takes an iterator, while `upload_object` takes a file path.

The reason I didn't implement `upload_object_via_stream` is because Backblaze API doesn't support streaming / multi part uploads. This means we need to buffer the whole file content in memory which is not ideal (especially for large files) and implementing `upload_object_via_stream` doesn't provide much value besides convenience if you already have an iterator object.

Having said that, I'm fine with implementing `upload_object_via_stream` for convenience, we just need to document and make it clear that this method does not perform streaming upload and that the whole file is read and buffered in memory.

Kami replied:
Edit: I see you also implemented `upload_object_via_stream` below. Adding `iterator` argument to this method breaks the standard API so one thing we could do is move common functionality into a new "private" method and then utilize this private method in `upload_object` and `upload_object_via_stream`.

jshridha replied:
@Kami I just noticed backblaze has a sneak peak of their new api method - b2_upload_part

Would it be better to implement [this](https://www.backblaze.com/b2/docs/b2_upload_part.html) as the upload_object_via_stream method?

jshridha replied:
I just tried to implement the large file upload and it is not yet supported by backblaze. I'm going to go forward with this temporary `upload_object_via_stream` method making the changes you suggested.

jshridha committed:
Fix for files with 0xe8 byte in file and python2

Kami commented:
Sorry for the delay and thanks for the contribution.

Kami commented on libcloud/storage/drivers/backblaze_b2.py:
```diff
@@ -301,8 +325,13 @@ def upload_object(self, file_path, container, object_name, extra=None,
         # don't support that
-        with open(file_path, 'rb') as fp:
-            iterator = iter(fp)
+        object_name = object_name.replace('\\', '/')
```
Might be a good idea to put this into a utility method - `escape/sanitize_object_name` or similar.

jshridha committed:
Refactor code to better adhere to the API requirements

Kami commented:
@jshridha Thanks.

Next time please ping us when you push updates since we don't get notified when you push new changes :)

Kami commented:
I've pushed some minor fixes / improvements and merged the patch into trunk (I also verified the patch is working).

Thanks!

jshridha commented:
@Kami Will do, I didn't realize that new pushes didn't result in notifications. Also, thanks for the great review!
